src/cunumeric/index/advanced_indexing.cc

```diff
@@ -22,74 +22,96 @@ namespace cunumeric {
 using namespace Legion;
 using namespace legate;
 
-template <LegateTypeCode CODE, int DIM1, int DIM2>
-struct AdvancedIndexingImplBody<VariantKind::CPU, CODE, DIM1, DIM2> {
+template <LegateTypeCode CODE, int DIM, typename OUT_TYPE>
+struct AdvancedIndexingImplBody<VariantKind::CPU, CODE, DIM, OUT_TYPE> {
   using VAL = legate_type_of<CODE>;
 
-  void compute_output(Buffer<VAL>& out,
-                      const AccessorRO<VAL, DIM1>& input,
-                      const AccessorRO<bool, DIM2>& index,
-                      const Pitches<DIM1 - 1>& pitches_input,
-                      const Rect<DIM1>& rect_input,
-                      const Pitches<DIM2 - 1>& pitches_index,
-                      const Rect<DIM2>& rect_index,
-                      const size_t volume) const
+  void compute_output(Buffer<VAL, DIM>& out,
+                      const AccessorRO<VAL, DIM>& input,
+                      const AccessorRO<bool, DIM>& index,
+                      const Pitches<DIM - 1>& pitches,
+                      const Rect<DIM>& rect,
+                      const int volume,
+                      const int key_dim,
+                      const int skip_size) const
   {
-    int64_t out_idx = 0;
+    size_t out_idx = 0;
     for (size_t idx = 0; idx < volume; ++idx) {
-      auto p = pitches_index.unflatten(idx, rect_index.lo);
+      auto p = pitches.unflatten(idx, rect.lo);
       if (index[p] == true) {
-        auto p_input = pitches_input.unflatten(idx, rect_input.lo);
-        out[out_idx] = input[p_input];
-        out_idx++;
+        Point<DIM> out_p;
+        out_p[0] = out_idx;
+        for (size_t i = 0; i < DIM - key_dim; i++) {
+          size_t j     = key_dim + i;
+          out_p[i + 1] = p[j];
+        }
+        for (size_t i = DIM - key_dim + 1; i < DIM; i++) out_p[i] = 0;
+        out[out_p] = input[p];
+        if ((idx + 1) % skip_size == 0) out_idx++;
       }
     }
   }
 
-  void compute_output(Buffer<Point<DIM1>>& out,
-                      const AccessorRO<VAL, DIM1>&,
-                      const AccessorRO<bool, DIM2>& index,
-                      const Pitches<DIM1 - 1>& pitches_input,
-                      const Rect<DIM1>& rect_input,
-                      const Pitches<DIM2 - 1>& pitches_index,
-                      const Rect<DIM2>& rect_index,
-                      const size_t volume) const
+  void compute_output(Buffer<Point<DIM>, DIM>& out,
```

Comment:
I think you can combine this overloaded function with the previous one with some template foo. Unless I'm mistaken the two only differ in the assignments and the rest is the same.